kyo-core/shared/src/main/scala/kyo/Clock.scala

```diff
@@ -104,21 +128,158 @@ object Clock:
+                    let(Clock(shifted))(v)
+                }
+            }
+    end withTimeShift
+
+    /** Interface for controlling time in a test environment.
+      *
+      * WARNING: TimeControl is not thread-safe. All operations should be performed sequentially to avoid race conditions and unexpected
+      * behavior.
+      */
+    trait TimeControl:
+        /** Sets the current time to a specific instant.
+          *
+          * @param now
+          *   The instant to set the current time to
+          * @return
+          *   Unit effect that updates the current time
+          */
+        def set(now: Instant): Unit < IO
+
+        /** Advances the current time by the specified duration.
+          *
+          * @param duration
+          *   The duration to advance time by
+          * @return
+          *   Unit effect that advances the current time
+          */
+        def advance(duration: Duration): Unit < IO
+    end TimeControl
+
+    /** Runs an effect with a controlled Clock that allows manual time manipulation. This is primarily intended for testing scenarios where
+      * precise control over time progression is needed.
+      *
+      * Note: TimeControl is not thread-safe. Operations on TimeControl should be performed sequentially within the same fiber.
+      *
+      * @param f
+      *   A function that takes a TimeControl and returns an effect
+      * @return
+      *   The result of running the effect with controlled time
+      */
+    def withTimeControl[A, S](f: TimeControl => A < S)(using Frame): A < (IO & S) =
+        IO.Unsafe {
+            val controlled =
+                new Unsafe with TimeControl:
+                    @volatile var current = Instant.Epoch
+
+                    case class Task(deadline: Instant) extends IOPromise[Nothing, Unit] with Delayed:
+                        def getDelay(unit: TimeUnit): Long =
+                            (deadline - current).to(unit)
+                        def compareTo(other: Delayed): Int =
+                            deadline.toJava.compareTo(other.asInstanceOf[Task].deadline.toJava)
+                    end Task
+
+                    val queue = new DelayQueue[Task]
+
+                    def now()(using AllowUnsafe) = current
+
+                    def sleep(duration: Duration): Fiber.Unsafe[Nothing, Unit] =
+                        val task = new Task(current + duration)
+                        queue.add(task)
+                        Promise.Unsafe.fromIOPromise(task)
+                    end sleep
+
+                    def set(now: Instant) =
+                        IO {
+                            current = now
+                            tick()
+                        }
+
+                    def advance(duration: Duration) =
+                        IO {
+                            current = current + duration
+                            tick()
+                        }
+
+                    def tick(): Unit =
```

Comment:
When the user changes the time, this method is called to execute the delayed tasks if it's time. The `DelayQueue` returns `null` in case there are no items ready for execution